lib/entice/logic/casting.ex

```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
   alias Entice.Logic.Casting
   alias Entice.Logic.Vitals.Energy
   alias Entice.Entity
-  alias Entice.Entity.Coordination
+  use Pipe
 
 
   defstruct(
-    casting_timer: nil,
+    casting_timer: nil,  #Why not call this cast_timer if skill has a cast_time ? or vice versa
     after_cast_timer: nil,
     recharge_timers: %{})
 
-
   def register(entity_id),
-  do: Entity.put_behaviour(entity_id, Casting.Behaviour, [])
-
+  do: Entity.put_behaviour(entity_id, Casting.Behaviour, %Casting{})
 
   def unregister(entity_id),
-  do: Entity.remove_behaviour(entity_id, Casting.Behaviour)
-
+  do: Entity.remove_behaviour(entity_id, Casting)
 
   @doc "Deals with timing and thus might fail. Should be called by the Skillbar"
   def cast_skill(entity, skill) when is_atom(skill),
-  do: entity |> Entity.call_behaviour({:skill_cast_start, %{target: nil, skill: skill}})
+  do: entity |> Entity.call_behaviour(Casting.Behaviour, {:cast_start, nil, nil, %{target: nil, skill: skill}})
 
   def cast_skill(entity, target, skill) when is_atom(skill),
-  do: entity |> Entity.call_behaviour({:skill_cast_start, %{target: target, skill: skill}})
+  do: Entity.call_behaviour(entity, Casting.Behaviour, {:cast_start, nil, nil, %{target: target, skill: skill}})
 
 
   defmodule Behaviour do
     use Entice.Entity.Behaviour
 
-    def handle_call(
-        {:skill_cast_start, %{target: target, skill: skill}},
-        %Entity{attributes: %{Energy => %Energy{mana: mana}}} = entity) do
-
-      case mana - skill.energy_cost do
-        new_mana when new_mana < 0 -> {:ok, {:error, :not_enough_energy}, entity}
-        new_mana ->
-          case skill.effect_cast_start(entity, target) do
-            {:error, reason} -> {:ok, {:error, reason}, entity}
-            {:ok, %Entity{} = entity, %Entity{} = target} ->
-              target.id |> Entity.attribute_transaction(&Map.merge(&1, target.attributes))
-              {:ok, :ok, entity}
-            result -> raise "Incorrect skill casting result:\nSkill: #{skill.id}\nResult: #{result}"
-          end
+    def init(entity, %Casting{} = casting),
+    do: {:ok, entity |> put_attribute(casting)}
+
+    def terminate(_reason, entity),
+    do: {:ok, entity |> remove_attribute(Casting)}
+
+    def handle_call({:cast_start, cast_callback, recharge_callback, %{target: _target, skill: skill}}, %Entity{attributes: %{
```

Comment:
lets fail gracefully instead of crashing by implementing a catch all default handle_call below this method - it might well be that the entity doesn't have energy